include/unicore-mx/nrf/51/gpio.h

```diff
@@ -0,0 +1,203 @@
+
+/* GPIO Tasks and Events (GPIOTE) */
+#define GPIO_TASK_OUT(n)			MMIO32(GPIOTE_BASE + 0x4 * (n))
+#define GPIO_TASK_OUT0			GPIO_TASK_OUT(0)
+#define GPIO_TASK_OUT1			GPIO_TASK_OUT(1)
+#define GPIO_TASK_OUT2			GPIO_TASK_OUT(2)
+#define GPIO_TASK_OUT3			GPIO_TASK_OUT(3)
+
+#define GPIO_EVENT_IN(n)			MMIO32(GPIOTE_BASE + 0x100 + 0x4 * (n))
+#define GPIO_EVENT_IN0			GPIO_EVENT_IN(0)
+#define GPIO_EVENT_IN1			GPIO_EVENT_IN(1)
+#define GPIO_EVENT_IN2			GPIO_EVENT_IN(2)
+#define GPIO_EVENT_IN3			GPIO_EVENT_IN(3)
+
+#define GPIO_EVENT_PORT			MMIO32(GPIOTE_BASE + 0x17C)
+
+#define GPIO_INTEN			MMIO32(GPIOTE_BASE + 0x300)
+#define GPIO_INTENSET			MMIO32(GPIOTE_BASE + 0x304)
+#define GPIO_INTENCLR			MMIO32(GPIOTE_BASE + 0x308)
+
+#define GPIO_TE_CONFIG(n)			MMIO32(GPIOTE_BASE + 0x510 + 0x4 * (n))
+#define GPIO_TE_CONFIG0			GPIO_TE_CONFIG(0)
+#define GPIO_TE_CONFIG1			GPIO_TE_CONFIG(1)
+#define GPIO_TE_CONFIG2			GPIO_TE_CONFIG(2)
+#define GPIO_TE_CONFIG3			GPIO_TE_CONFIG(3)
+
+/* Register Details */
+#define GPIO_INTEN_IN(n)			(1 << (n))
+#define GPIO_INTEN_IN0			(1 << 0)
+#define GPIO_INTEN_IN1			(1 << 1)
+#define GPIO_INTEN_IN2			(1 << 2)
+#define GPIO_INTEN_IN3			(1 << 3)
+
+#define GPIO_INTEN_PORT			(1 << 31)
+
+#define GPIO_TE_CONFIG_MODE_SHIFT			(0)
+#define GPIO_TE_CONFIG_MODE_MASK			(3 << GPIO_TE_CONFIG_MODE_SHIFT)
+#define GPIO_TE_CONFIG_MODE_MASKED(V)			(((V) << GPIO_TE_CONFIG_MODE_SHIFT) & GPIO_TE_CONFIG_MODE_MASK)
+
+#define GPIO_TE_CONFIG_PSEL_SHIFT			(8)
+#define GPIO_TE_CONFIG_PSEL_MASK			(0x1f << GPIO_TE_CONFIG_PSEL_SHIFT)
+#define GPIO_TE_CONFIG_PSEL_MASKED(V)			(((V) << GPIO_TE_CONFIG_PSEL_SHIFT) & GPIO_TE_CONFIG_PSEL_MASK)
+
+#define GPIO_TE_CONFIG_POLARITY_SHIFT			(16)
+#define GPIO_TE_CONFIG_POLARITY_MASK			(3 << GPIO_TE_CONFIG_POLARITY_SHIFT)
+#define GPIO_TE_CONFIG_POLARITY_MASKED(V)			(((V) << GPIO_TE_CONFIG_POLARITY_SHIFT) & GPIO_TE_CONFIG_POLARITY_MASK)
+
+#define GPIO_TE_CONFIG_OUTINIT			(1 << 20)
+
+enum gpio_dir {
+	GPIO_DIR_INPUT,
+	GPIO_DIR_OUTPUT,
+};
+
+enum gpio_input {
+	GPIO_INPUT_CONNECT,
+	GPIO_INPUT_DISCONNECT,
+};
+
+enum gpio_pull {
+	GPIO_PULL_NONE,
+	GPIO_PULL_DOWN,
+	GPIO_PULL_UP = 3,
+};
+
+enum gpio_drive {
+	GPIO_DRIVE_S0S1,
+	GPIO_DRIVE_H0S1,
+	GPIO_DRIVE_S0H1,
+	GPIO_DRIVE_H0H1,
+	GPIO_DRIVE_D0S1,
+	GPIO_DRIVE_D0H1,
+	GPIO_DRIVE_S0D1,
+	GPIO_DRIVE_H0D1,
+};
+
+enum gpio_te_mode {
+	GPIO_TE_MODE_DISABLED,
+	GPIO_TE_MODE_EVENT,
+	GPIO_TE_MODE_TASK = 3,
+};
+
+enum gpio_te_polarity {
+	GPIO_TE_POLARITY_NONE,
+	GPIO_TE_POLARITY_LO_TO_HI,
+	GPIO_TE_POLARITY_HI_TO_LO,
+	GPIO_TE_POLARITY_TOGGLE,
+};
+
+enum gpio_te_outinit {
+	GPIO_TE_OUTINIT_LOW,
+	GPIO_TE_OUTINIT_HIGH = GPIO_TE_CONFIG_OUTINIT,
+};
+
+BEGIN_DECLS
+
+void gpio_set(uint32_t gpios);
+void gpio_clear(uint32_t gpios);
+void gpio_toggle(uint32_t gpios);
+uint32_t gpio_get(uint32_t gpios);
+
+void gpio_set_direction(uint32_t gpios, enum gpio_dir dir);
+void gpio_set_pull(uint32_t gpios, enum gpio_pull pull);
+
+void gpio_configure_task(uint8_t task_num,
```

Comment:
Can we reuse function prototypes from other platforms such as  gpio_mode_setup/gpio_set_output_options/gpio_set_af, and keep arguments "generic" as uint32_t? This would facilitate our task of having the GPIO API as much unified as possible. Enums above should then be replaced by #defines